test/quic_tserver_test.c

```diff
@@ -0,0 +1,221 @@
+
+static int is_want(SSL *s, int ret)
+{
+    int ec = SSL_get_error(s, ret);
+    return ec == SSL_ERROR_WANT_READ || ec == SSL_ERROR_WANT_WRITE;
+}
+
+static int test_tserver(void)
+{
+    int testresult = 0, ret;
+    int s_fd = -1, c_fd = -1;
+    BIO *s_net_bio = NULL, *s_net_bio_own = NULL;
+    BIO *c_net_bio = NULL, *c_net_bio_own = NULL;
+    QUIC_TSERVER_ARGS tserver_args = {0};
+    QUIC_TSERVER *tserver = NULL;
+    BIO_ADDR *s_addr_ = NULL;
+    struct in_addr ina = {0};
+    union BIO_sock_info_u s_info = {0};
+    SSL_CTX *c_ctx = NULL;
+    SSL *c_ssl = NULL;
+    short port = 8186;
+    int c_connected = 0, c_write_done = 0, c_begin_read = 0;
+    size_t l = 0, s_total_read = 0, s_total_written = 0, c_total_read = 0;
+    int s_begin_write = 0;
+    OSSL_TIME start_time;
+
+    ina.s_addr = htonl(0x7f000001UL);
+
+    /* Setup test server. */
+    s_fd = BIO_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP, 0);
+    if (!TEST_int_ge(s_fd, 0))
+        goto err;
+
+    if (!TEST_true(BIO_socket_nbio(s_fd, 1)))
+        goto err;
+
+    if (!TEST_ptr(s_addr_ = BIO_ADDR_new()))
+        goto err;
+
+    if (!TEST_true(BIO_ADDR_rawmake(s_addr_, AF_INET, &ina, sizeof(ina),
+                                    htons(port))))
+        goto err;
+
+    if (!TEST_true(BIO_bind(s_fd, s_addr_, 0)))
+        goto err;
+
+    s_info.addr = s_addr_;
+    if (!TEST_true(BIO_sock_info(s_fd, BIO_SOCK_INFO_ADDRESS, &s_info)))
+        goto err;
+
+    if (!TEST_int_gt(BIO_ADDR_rawport(s_addr_), 0))
+        goto err;
+
+    if (!TEST_ptr(s_net_bio = s_net_bio_own = BIO_new_dgram(s_fd, 0)))
+        goto err;
+
+    if (!BIO_up_ref(s_net_bio))
+        goto err;
+
+    tserver_args.net_rbio = s_net_bio;
+    tserver_args.net_wbio = s_net_bio;
+
+    if (!TEST_ptr(tserver = ossl_quic_tserver_new(&tserver_args))) {
+        BIO_free(s_net_bio);
+        goto err;
+    }
+
+    s_net_bio_own = NULL;
+
+    /* Setup test client. */
+    c_fd = BIO_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP, 0);
+    if (!TEST_int_ge(c_fd, 0))
+        goto err;
+
+    if (!TEST_true(BIO_socket_nbio(c_fd, 1)))
+        goto err;
+
+    if (!TEST_ptr(c_net_bio = c_net_bio_own = BIO_new_dgram(c_fd, 0)))
+        goto err;
+
+    if (!BIO_dgram_set_peer(c_net_bio, s_addr_))
+        goto err;
+
+    if (!TEST_ptr(c_ctx = SSL_CTX_new(OSSL_QUIC_client_method())))
+        goto err;
+
+    if (!TEST_ptr(c_ssl = SSL_new(c_ctx)))
+        goto err;
+
+    /* Takes ownership of our reference to the BIO. */
+    SSL_set0_rbio(c_ssl, c_net_bio);
+
+    /* Get another reference to be transferred in the SSL_set0_wbio call. */
+    if (!TEST_true(BIO_up_ref(c_net_bio))) {
+        c_net_bio_own = NULL; /* SSL_free will free the first reference. */
+        goto err;
+    }
+
+    SSL_set0_wbio(c_ssl, c_net_bio);
+    c_net_bio_own = NULL;
+
+    if (!TEST_true(SSL_set_blocking_mode(c_ssl, 0)))
+        goto err;
+
+    start_time = ossl_time_now();
+
+    for (;;) {
```

Comment:
It might be an idea to have a maxloops counter or similar here, so that we always eventually break and don't hang in the test in the event of an unexpected failure to connect.

Reply:
This timeout is implemented chronologically below?

Reply:
> This timeout is implemented chronologically below?

Ah! Right. Yes - that's good enough.